epochStart/metachain/systemSCs.go

```diff
@@ -327,9 +333,134 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
+		if err != nil {
+			return err
+		}
+
+		err = s.selectNodesFromAuctionList(validatorInfos, randomness)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *systemSCProcessor) selectNodesFromAuctionList(validatorInfos map[uint32][]*state.ValidatorInfo, randomness []byte) error {
+	auctionList, noOfValidators := getAuctionListAndNoOfValidators(validatorInfos)
+	err := s.sortAuctionList(auctionList, randomness)
+	if err != nil {
+		return err
+	}
+
+	auctionListSize := uint32(len(auctionList))
+	noOfAvailableNodeSlots := core.MinUint32(auctionListSize, s.maxNodes-noOfValidators)
+	s.displayAuctionList(auctionList, noOfAvailableNodeSlots)
+
+	for i := uint32(0); i < noOfAvailableNodeSlots; i++ {
+		auctionList[i].List = string(common.NewList)
+	}
+
+	return nil
+}
+
+func getAuctionListAndNoOfValidators(validatorInfos map[uint32][]*state.ValidatorInfo) ([]*state.ValidatorInfo, uint32) {
+	auctionList := make([]*state.ValidatorInfo, 0)
+	noOfValidators := uint32(0)
+
+	for _, validatorsInShard := range validatorInfos {
+		for _, validator := range validatorsInShard {
+			if validator.List == string(common.AuctionList) {
+				auctionList = append(auctionList, validator)
+			} else if isValidator(validator) {
+				noOfValidators++
+			}
+		}
+	}
+
+	return auctionList, noOfValidators
+}
+
+func (s *systemSCProcessor) sortAuctionList(auctionList []*state.ValidatorInfo, randomness []byte) error {
+	errors := make([]error, 0)
+
+	sort.SliceStable(auctionList, func(i, j int) bool {
+		pubKey1 := auctionList[i].PublicKey
+		pubKey2 := auctionList[j].PublicKey
+
+		nodeTopUpPubKey1, err := s.stakingDataProvider.GetNodeStakedTopUp(pubKey1)
+		if err != nil {
+			errors = append(errors, err)
+			log.Debug(fmt.Sprintf("%v when trying to get top up per node for %s", err, hex.EncodeToString(pubKey1)))
+		}
+
+		nodeTopUpPubKey2, err := s.stakingDataProvider.GetNodeStakedTopUp(pubKey2)
+		if err != nil {
+			errors = append(errors, err)
+			log.Debug(fmt.Sprintf("%v when trying to get top up per node for %s", err, hex.EncodeToString(pubKey2)))
+		}
+
+		if nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) == 0 {
+			return compareByXORWithRandomness(pubKey1, pubKey2, randomness)
+		}
+
+		return nodeTopUpPubKey1.Cmp(nodeTopUpPubKey2) == 1
+	})
+
+	if len(errors) > 0 {
+		return fmt.Errorf("%w; last known error %v", epochStart.ErrSortAuctionList, errors[len(errors)-1])
+	}
 	return nil
 }
 
+func compareByXORWithRandomness(pubKey1, pubKey2, randomness []byte) bool {
+	key1Xor := make([]byte, len(randomness))
+	key2Xor := make([]byte, len(randomness))
+
+	for idx := range randomness {
+		key1Xor[idx] = pubKey1[idx] ^ randomness[idx]
+		key2Xor[idx] = pubKey2[idx] ^ randomness[idx]
+	}
+
+	return bytes.Compare(key1Xor, key2Xor) == 1
+}
+
+func (s *systemSCProcessor) displayAuctionList(auctionList []*state.ValidatorInfo, noOfSelectedNodes uint32) {
+	tableHeader := []string{"Owner", "Registered key", "TopUp per node"}
```

Comment:
suggestion to add this here:
```GO
	if log.GetLevel() > logger.LogDebug{
		return
	}
```
this will help a lot of useless processing in case the log level is higher than the Debug level

Reply:
Very good suggestion. Applied it ! :+1: